### DISCUSSION TITLE
button in datagrid ,how to bind command

### DISCUSSION THREAD

**Discussion by Cohn21** [AvaloniaUI/Avalonia#12821] [Q&A] [answered]
`<DataGrid
            IsReadOnly="True"
            ItemsSource="{Binding Lists}">
  <DataGrid.Columns>
 <DataGridTemplateColumn Width="*">
                    <DataGridTemplateColumn.CellTemplate>
                        <DataTemplate>
                            <Button Command="{Binding DataContext.CheckCommand,RelativeSource={RelativeSource AncestorType={x:Type DataGrid}}}"/>
                        </DataTemplate>
                    </DataGridTemplateColumn.CellTemplate>
                </DataGridTemplateColumn>
            </DataGrid.Columns>
        </DataGrid>`
CheckCommand in window.datacontext,not in Lists,can not bind
this has error

**Comment by timunie** [comment]
Give your DataGrid a name and bind to named control. {binding #myDataGrid....}

**Reply by Cohn21** [reply]
CheckCommand in window's viewmodel,like   
`<Window>
 <Design.DataContext>
        <vm:MainWindowViewModel />
    </Design.DataContext>
    <DataGrid ItemsSourcc={Binding Lists}>
     <DataGridTemplateColumn>
                    <DataGridTemplateColumn.CellTemplate>
                        <DataTemplate>
                           <Button Command="{Binding MainWindowViewModel.CheckCommand"}/>
                        </DataTemplate>
                    </DataGridTemplateColumn.CellTemplate>
                </DataGridTemplateColumn>
    </DataGrid>
</Window>`

**Reply by timunie** [reply, marked as answer]
No TemplateColumn is binding to your Item's datacontext. So you have to bind to parents data, or you have to pass parent ViewModel to ItemViewModel. 

```diff
<Window> 
  <Design.DataContext>
     <vm:MainWindowViewModel /> 
  </Design.DataContext> 
! <DataGrid x:Name="myDataGrid"
            ItemsSourcc={Binding Lists}> 
    <DataGridTemplateColumn> 
      <DataGridTemplateColumn.CellTemplate> 
        <DataTemplate> 
          <!-- If you have CompiledBindings enabled, you need to cast the DataContext to MainViewModel -->
!         <Button Command="{Binding #myDataGrid.((vm:MainWindowViewModel)DataContext).CheckCommand"}/> 
        </DataTemplate> 
      </DataGridTemplateColumn.CellTemplate> 
    </DataGridTemplateColumn> 
  </DataGrid> 
</Window>
```

Note: I copied your snippet and only changed the highlighted columns. Snippet is not complete and has typos, so don't just copy paste it.

**Reply by Cohn21** [reply]
thanks,work Ok

**Comment by pyking-wang** [comment]
Jump from the current data context (i.e. the context of a single data row) to the data context of the parent DataGrid control.
```
<DataTemplate>
	<Button Content=”Test“ Command="{Binding $parent[DataGrid].DataContext.ContextMenuCommand}"/>
</DataTemplate>
```
